firedrake/petsc.py

```diff
@@ -249,3 +251,46 @@ def inserted_options(self):
         finally:
             for k in self.to_delete:
                 del self.options_object[self.options_prefix + k]
+
+
+def _extract_comm(obj):
+    comm = None
+    if isinstance(obj, (PETSc.Comm, mpi.MPI.Comm)):
+        try:
+            if mpi.is_pyop2_comm(obj):
+                comm = obj
+            else:
+                internal_comm = obj.Get_attr(mpi.innercomm_keyval)
+                if internal_comm is None:
+                    comm = obj
+                else:
+                    comm = internal_comm
```

Comment:
I have decided that I don't like the policy of "only ever clean up the internal communicator". Consider the following:
```py
mesh = UnitSquareMesh(...)
vec = PETSc.Vec().create(mesh.comm)
garbage_cleanup(mesh.comm)  # or garbage_cleanup(mesh)
```
Since `vec` is associated with the public mesh communicator (`mesh.comm` instead of `mesh._comm`) then it will never be cleaned up by `garbage_cleanup` since that will always act on `mesh._comm`.

I propose that `garbage_cleanup` look like the following:
```py
def garbage_cleanup(obj):
    if isinstance(obj, ParallelObject):
        PETSc.garbage_cleanup(obj._comm)
    else:
        PETSc.garbage_cleanup(obj)
```

In other words, if one passes a parallel Firedrake (or PyOP2) object then call the cleanup on the associated internal communicator. If a communicator is passed instead then clean up *on that communicator* - don't extract the private one.

I added `ParallelObject` here because I would really like a mixin class encapsulating the `.comm`, `._comm`, `__del__`, `mpi.internal_comm` stuff repeated all over the place. This isn't strictly necessary though and the `isinstance()` call could be replaced by `hasattr(obj, "_comm")`.

Reply:
Just to note, when I (many years ago) wrote the comm-duplication code in pyop2 to deliver "library" communicators, I mimiced the petsc setup (since at the time we thought people might use pyop2 without petsc).

So there are three communicators in play in this example:

1. the user communicator (passed in to `Mesh`), `mesh.comm`
2. the internal PyOP2 communicator `mesh._comm`
3. the internal PETSc communicator `vec.comm` (which is not `mesh.comm`, but is stashed inside it), see https://petsc.org/release/manualpages/Sys/PetscCommDuplicate/

When you do `VecCreate(user_comm, &vec)` internally petsc does `PetscCommDuplicate(user_comm, &petsc_comm)` and then `PetscObjectComm((PetscObject)vec)` returns `petsc_comm`. It used to be that Firedrake/PyOP2 did similar, so `mesh = Mesh(..., comm=user_comm)` did `firedrake_comm = dup_comm(user_comm)` and then `mesh.comm` would be `firedrake_comm` (not `user_comm`).

I think all of the garbage lives on the internal PETSc communicator, but petsc doesn't know about the connection between `mesh._comm` and `mesh.comm` so there will be two petsc communicators and not all the garbage will be collected.

I'm sure there are good reasons why `mesh._comm` is not just the communicator obtained from `PetscCommDuplicate` but I can't remember them. OTOH, if that were the case, would that make things simpler? `P etscGarbageCleanup` always cleans up on the internal petsc communicator, so as long as `mesh._comm` and `mesh.comm` have the same internal petsc communicator it doesn't matter if you're collecting on the public or he private one.

Reply:
> When you do `VecCreate(user_comm, &vec)` internally petsc does `PetscCommDuplicate(user_comm, &petsc_comm)` and then `PetscObjectComm((PetscObject)vec)` returns `petsc_comm`. It used to be that Firedrake/PyOP2 did similar, so `mesh = Mesh(..., comm=user_comm)` did `firedrake_comm = dup_comm(user_comm)` and then `mesh.comm` would be `firedrake_comm` (not `user_comm`).

I don't especially like the pattern of `mesh = Mesh(..., comm=mycomm); mesh.comm != mycomm`.

> I'm sure there are good reasons why `mesh._comm` is not just the communicator obtained from `PetscCommDuplicate` but I can't remember them. OTOH, if that were the case, would that make things simpler? `P etscGarbageCleanup` always cleans up on the internal petsc communicator, so as long as `mesh._comm` and `mesh.comm` have the same internal petsc communicator it doesn't matter if you're collecting on the public or he private one.

I think I'm in favour of dropping PyOP2-specific comms so our notions of public and private comms could match PETSc's. The majority of MPI calls than happen in PyOP2 are done via PETSc anyway (e.g. halo exchanges with PetscSFs) so I don't see where the benefit is.

Reply:
I disagree with this assessment, I believe that the example user code you wrote above should look like:
```python
mesh = UnitSquareMesh(...)
vec = PETSc.Vec().create(mesh.comm)
PETSc.garbage_cleanup(mesh.comm)
```
As is now [documented in petsc4py](https://petsc.org/release/petsc4py/reference/petsc4py.PETSc.garbage_cleanup.html). If a user is doing petsc4py things they should be cleaning them up using petsc4py methods, if they are doing Firedrake things they should be cleaning them up using Firedrake methods (or indeed not cleaning them up, Firedrake is doing this periodically anyway!). I think the Firedrake cleanup function cleaning petsc4py objects off a user comm (neither item "belonging" to Firedrake) is confusing.

I agree with `mesh = Mesh(..., comm=mycomm); mesh.comm != mycomm` being a bad idea, but I'm not in favour of dropping internal comms. Having internal "library" communicators is useful, there is a good summary and example on [this rookieHPC page](https://rookiehpc.org/mpi/docs/mpi_comm_dup/index.html). Indeed I'm happy with the three tiers of communicator, user comm, Firedrake internal/PyOP2 comm and PETSc comm. Firedrake is a library (for the purposes of this discussion) and PyOP2 handles the duplicating the user comm, PETSc is a separate library and should also have its own duplicated comm, whether that be a user comm, or a Firedrake.PyOP2 comm.

Let me know your thoughts.

Reply:
Fine. But our docstring for `garbage_cleanup` should make this clear.

Reply:
> example on [this rookieHPC page](https://rookiehpc.org/mpi/docs/mpi_comm_dup/index.html).

ohh... that example makes this all make sense!